numpy/lib/arraypad.py

```diff
@@ -986,399 +136,191 @@ def _validate_lengths(narray, number_elements):
-    >>> np.pad(a, (2, 3), 'reflect')
-    array([3, 2, 1, 2, 3, 4, 5, 4, 3, 2])
-
-    >>> np.pad(a, (2, 3), 'reflect', reflect_type='odd')
-    array([-1,  0,  1,  2,  3,  4,  5,  6,  7,  8])
-
-    >>> np.pad(a, (2, 3), 'symmetric')
-    array([2, 1, 1, 2, 3, 4, 5, 5, 4, 3])
-
-    >>> np.pad(a, (2, 3), 'symmetric', reflect_type='odd')
-    array([0, 1, 1, 2, 3, 4, 5, 5, 6, 7])
-
-    >>> np.pad(a, (2, 3), 'wrap')
-    array([4, 5, 1, 2, 3, 4, 5, 1, 2, 3])
-
-    >>> def pad_with(vector, pad_width, iaxis, kwargs):
-    ...     pad_value = kwargs.get('padder', 10)
-    ...     vector[:pad_width[0]] = pad_value
-    ...     vector[-pad_width[1]:] = pad_value
-    ...     return vector
-    >>> a = np.arange(6)
-    >>> a = a.reshape((2, 3))
-    >>> np.pad(a, 2, pad_with)
-    array([[10, 10, 10, 10, 10, 10, 10],
-           [10, 10, 10, 10, 10, 10, 10],
-           [10, 10,  0,  1,  2, 10, 10],
-           [10, 10,  3,  4,  5, 10, 10],
-           [10, 10, 10, 10, 10, 10, 10],
-           [10, 10, 10, 10, 10, 10, 10]])
-    >>> np.pad(a, 2, pad_with, padder=100)
-    array([[100, 100, 100, 100, 100, 100, 100],
-           [100, 100, 100, 100, 100, 100, 100],
-           [100, 100,   0,   1,   2, 100, 100],
-           [100, 100,   3,   4,   5, 100, 100],
-           [100, 100, 100, 100, 100, 100, 100],
-           [100, 100, 100, 100, 100, 100, 100]])
+    padded : ndarray
+        Larger array with undefined values in padded areas.
     """
+    # Allocate grown array
+    new_shape = tuple(
+        size + left + right
+        for size, (left, right) in zip(arr.shape, pad_widths)
+    )
+    padded = np.empty(new_shape, dtype=arr.dtype)
+
+    # Copy old array into correct space
+    old_area = tuple(
+        slice(left, left + size)
+        for size, (left, right) in zip(arr.shape, pad_widths)
+    )
+    padded[old_area] = arr
+
+    return padded
+
+
+def _set_generic(arr, axis, pad_index, values):
+    pad_area = _slice_pad_area(arr.shape, axis, pad_index)
+    arr[pad_area] = values
+
+
+def _set_edge(arr, axis, pad_index):
+    edge_slice = _slice_column(arr.shape, axis, pad_index)
+    edge_arr = arr[edge_slice].repeat(abs(pad_index), axis=axis)
+    _set_generic(arr, axis, pad_index, edge_arr)
+
+
+def _set_linear_ramp(arr, axis, pad_index, end_value):
+    pad_shape = arr.shape[:axis] + (abs(pad_index),) + arr.shape[(axis + 1):]
+    reverse = True if 0 <= pad_index else False
+    linear_ramp = _arange_ndarray(tuple(pad_shape), axis, reverse)
+
+    edge_slice = _slice_column(arr.shape, axis, pad_index)
+    edge_arr = arr[edge_slice].repeat(abs(pad_index), axis=axis)
+
+    slope = (end_value - edge_arr) / float(abs(pad_index))
+    linear_ramp = linear_ramp * slope
+    linear_ramp += edge_arr
+    _round_if_needed(linear_ramp, arr.dtype)
+
+    _set_generic(arr, axis, pad_index,
+                 linear_ramp.astype(arr.dtype, copy=False))
+
+
+def pad(array, pad_width, mode, stat_length=None, constant_values=0,
```

Comment:
This changes `constant_values` to no longer be keyword-only, which I think is undesirable.

Reply:
I guess the Python 3 syntax using the `*` to force keyword-only arguments isn't an option because Python 2.7 must be supported?
I see the benefits of keyword-only arguments but in this case I thought the handling with `allowedkwargs` and `kwdefauts` was awkward.

But I'll change it back for now.